// awal/dist/ipcClient.js
import * as fs from 'fs';
import * as path from 'path';
import { randomUUID } from 'crypto';
import { getPrimaryProcessPid } from './utils/processCheck.js';
import { ensureServerRunning } from './utils/serverManager.js';
import { IpcTimeoutError, IpcBridgeError } from './utils/errors.js';
// Set process title for security validation in WindowBridge
process.title = 'awal-cli';
// Callback for when wallet is starting (set by CLI for spinner updates)
let onWalletStartingCallback = null;
let onInstallingCallback = null;
/**
 * Sets a callback to be called when the wallet is being started
 */
export function setOnWalletStarting(callback) {
    onWalletStartingCallback = callback;
}
/**
 * Sets a callback to be called when the server is being installed
 */
export function setOnInstalling(callback) {
    onInstallingCallback = callback;
}
// IPC directories matching WindowBridge.ts
const ipcDir = '/tmp/payments-mcp-ui-bridge';
const requestsDir = path.join(ipcDir, 'requests');
const responsesDir = path.join(ipcDir, 'responses');
/**
 * Ensures IPC directories exist with secure permissions
 */
function ensureIpcDirs() {
    try {
        if (!fs.existsSync(ipcDir))
            fs.mkdirSync(ipcDir, { mode: 0o700 });
        if (!fs.existsSync(requestsDir))
            fs.mkdirSync(requestsDir, { mode: 0o700 });
        if (!fs.existsSync(responsesDir))
            fs.mkdirSync(responsesDir, { mode: 0o700 });
    }
    catch (error) {
        throw new IpcBridgeError(`Failed to create IPC directories: ${error instanceof Error ? error.message : String(error)}`);
    }
}
/**
 * Sends an IPC request to the primary Electron process and waits for response
 * Auto-starts the wallet if not running.
 */
export async function sendIpcRequest(channel, data) {
    // Check if wallet is running, start if not
    if (getPrimaryProcessPid() === null) {
        const started = await ensureServerRunning(onInstallingCallback ?? undefined, onWalletStartingCallback ?? undefined);
        if (!started) {
            throw new IpcBridgeError('Failed to start wallet. Please start it manually.');
        }
    }
    // Ensure IPC directories exist
    ensureIpcDirs();
    const requestId = randomUUID();
    const requestFile = path.join(requestsDir, `${requestId}.json`);
    const responseFile = path.join(responsesDir, `${requestId}.json`);
    try {
        // Write request to file with process information for validation
        const request = {
            id: requestId,
            channel,
            data,
            timestamp: Date.now(),
            pid: process.pid,
            processTitle: process.title,
        };
        // Write with secure permissions (600 - owner read/write only)
        fs.writeFileSync(requestFile, JSON.stringify(request, null, 2), { mode: 0o600 });
        // Wait for response file using file system watcher
        const maxWait = 30000; // 30 seconds
        await new Promise((resolve, reject) => {
            const timeout = setTimeout(() => {
                watcher.close();
                reject(new IpcTimeoutError(channel, maxWait));
            }, maxWait);
            // Check if file already exists
            if (fs.existsSync(responseFile)) {
                clearTimeout(timeout);
                resolve();
                return;
            }
            // Watch for file creation
            const watcher = fs.watch(responsesDir, (eventType, filename) => {
                if (filename === `${requestId}.json`) {
                    clearTimeout(timeout);
                    watcher.close();
                    resolve();
                }
            });
        });
        // Read and parse response
        const responseData = fs.readFileSync(responseFile, 'utf8');
        const response = JSON.parse(responseData);
        // Clean up response file
        try {
            fs.unlinkSync(responseFile);
        }
        catch {
            // Ignore cleanup errors
        }
        if (response.error) {
            const errorMsg = typeof response.error === 'string'
                ? response.error
                : JSON.stringify(response.error);
            throw new IpcBridgeError(errorMsg);
        }
        return response.result;
    }
    catch (error) {
        // Clean up response file on error
        try {
            if (fs.existsSync(responseFile))
                fs.unlinkSync(responseFile);
        }
        catch {
            // Ignore cleanup errors
        }
        // Re-throw our custom errors as-is
        if (error instanceof IpcTimeoutError ||
            error instanceof IpcBridgeError) {
            throw error;
        }
        throw new IpcBridgeError(error instanceof Error
            ? error.message
            : typeof error === 'object' && error !== null
                ? JSON.stringify(error)
                : String(error));
    }
}
