// awal/dist/formatters/output.js
import chalk from 'chalk';
/**
 * Formats data for JSON output
 */
export function formatJson(data) {
    return JSON.stringify(data, null, 2);
}
/**
 * Formats key-value pairs for text output
 */
export function formatKeyValue(data) {
    const lines = [];
    for (const [key, value] of Object.entries(data)) {
        const formattedValue = typeof value === 'object' ? JSON.stringify(value) : String(value);
        lines.push(`${chalk.bold(key)}: ${formattedValue}`);
    }
    return lines.join('\n');
}
/**
 * Formats an array of objects as a simple table
 */
export function formatTable(data, columns) {
    if (data.length === 0) {
        return 'No data';
    }
    // Calculate column widths
    const widths = columns.map((col) => {
        const headerWidth = col.header.length;
        const maxDataWidth = Math.max(...data.map((row) => {
            const value = row[col.key];
            return String(value ?? '').length;
        }));
        return col.width ?? Math.max(headerWidth, maxDataWidth);
    });
    // Build header
    const headerLine = columns.map((col, i) => col.header.padEnd(widths[i])).join('  ');
    // Build separator
    const separatorLine = widths.map((w) => '-'.repeat(w)).join('  ');
    // Build rows
    const rowLines = data.map((row) => columns
        .map((col, i) => {
        const value = row[col.key];
        const str = String(value ?? '');
        return str.length > widths[i] ? str.slice(0, widths[i] - 3) + '...' : str.padEnd(widths[i]);
    })
        .join('  '));
    return [chalk.bold(headerLine), separatorLine, ...rowLines].join('\n');
}
/**
 * Formats balance display
 */
export function formatBalance(formatted, token, chain) {
    const tokenInfo = token ? ` ${token}` : '';
    const chainInfo = chain ? ` (${chain})` : '';
    return `${chalk.green(formatted)}${tokenInfo}${chainInfo}`;
}
/**
 * Formats a success message
 */
export function formatSuccess(message) {
    return chalk.green(`✓ ${message}`);
}
/**
 * Formats a warning message
 */
export function formatWarning(message) {
    return chalk.yellow(`⚠ ${message}`);
}
/**
 * Formats an info message
 */
export function formatInfo(message) {
    return chalk.blue(`ℹ ${message}`);
}
/**
 * Formats USDC atomic units to human-readable format
 */
export function formatUsdcAmount(atomicUnits) {
    const units = typeof atomicUnits === 'string' ? parseInt(atomicUnits, 10) : atomicUnits;
    const dollars = units / 1_000_000;
    return `$${dollars.toFixed(2)} USDC`;
}
/**
 * Formats asset balance based on asset type and decimals.
 * USDC shows as $X.XX, ETH/WETH/SOL show with appropriate precision.
 */
export function formatAssetBalance(asset, raw, decimals) {
    const balance = Number(raw) / Math.pow(10, decimals);
    if (asset.toUpperCase() === 'USDC') {
        return `$${balance.toFixed(2)}`;
    }
    // For ETH/WETH/SOL, show more precision for small amounts
    if (balance === 0) {
        return '0.00';
    }
    if (balance < 0.0001) {
        return balance.toFixed(6);
    }
    return balance.toFixed(4);
}
