// awal/dist/utils/bazaarCache.js
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'fs';
import * as path from 'path';
import * as os from 'os';
import { Document } from '@langchain/core/documents';
import { BM25Retriever } from '@langchain/community/retrievers/bm25';
const DEFAULT_PAGE_SIZE = 100;
const STALE_HOURS = 12;
const CDP_FACILITATOR = {
    url: 'https://api.cdp.coinbase.com/platform/v2/x402/discovery/resources',
    pageSize: 1000,
};
const OPENX402_FACILITATOR = {
    url: 'https://facilitator.openx402.ai/discovery/resources',
    pageSize: DEFAULT_PAGE_SIZE,
};
const FACILITATORS = [CDP_FACILITATOR, OPENX402_FACILITATOR];
function getCacheDir() {
    return path.join(os.homedir(), '.config', 'awal', 'bazaar');
}
function getCachePath() {
    return path.join(getCacheDir(), 'resources.json');
}
async function fetchFromFacilitator(facilitator) {
    const all = [];
    let offset = 0;
    const pageSize = facilitator.pageSize || DEFAULT_PAGE_SIZE;
    let hasMore = true;
    while (hasMore) {
        const url = `${facilitator.url}?limit=${pageSize}&offset=${offset}`;
        const response = await fetch(url, {
            headers: { Accept: 'application/json' },
        });
        if (!response.ok) {
            if (response.status === 429 && all.length > 0) {
                break;
            }
            throw new Error(`Facilitator ${facilitator.url} returned ${response.status}: ${response.statusText}`);
        }
        const body = (await response.json());
        for (const item of body.items) {
            item.facilitator = facilitator.url;
        }
        all.push(...body.items);
        if (all.length >= body.pagination.total) {
            hasMore = false;
        }
        else {
            offset += pageSize;
        }
    }
    return all;
}
export async function fetchAllResources() {
    const results = await Promise.all(FACILITATORS.map(fetchFromFacilitator));
    return results.flat();
}
export function loadCache() {
    const cachePath = getCachePath();
    if (!existsSync(cachePath)) {
        return null;
    }
    try {
        const raw = readFileSync(cachePath, 'utf-8');
        return JSON.parse(raw);
    }
    catch {
        return null;
    }
}
export function saveCache(resources) {
    const dir = getCacheDir();
    if (!existsSync(dir)) {
        mkdirSync(dir, { recursive: true });
    }
    const cache = {
        fetchedAt: new Date().toISOString(),
        resources,
    };
    writeFileSync(getCachePath(), JSON.stringify(cache, null, 2), 'utf-8');
}
export function isCacheStale(cache) {
    const fetchedAt = new Date(cache.fetchedAt).getTime();
    const now = Date.now();
    return now - fetchedAt > STALE_HOURS * 60 * 60 * 1000;
}
export async function getResources(forceRefresh = false) {
    if (!forceRefresh) {
        const cache = loadCache();
        if (cache && !isCacheStale(cache)) {
            return cache.resources;
        }
    }
    const resources = await fetchAllResources();
    saveCache(resources);
    return resources;
}
function buildDocContent(r) {
    const parts = [r.resource];
    if (r.metadata?.description) {
        parts.push(r.metadata.description);
    }
    if (r.metadata?.input) {
        parts.push(JSON.stringify(r.metadata.input));
    }
    if (r.metadata?.output) {
        parts.push(JSON.stringify(r.metadata.output));
    }
    if (r.accepts) {
        for (const a of r.accepts) {
            if (a.description)
                parts.push(a.description);
            if (a.network)
                parts.push(a.network);
            if (a.scheme)
                parts.push(a.scheme);
            if (a.asset)
                parts.push(a.asset);
        }
    }
    if (r.type) {
        parts.push(r.type);
    }
    return parts.join(' ');
}
const GENERIC_DESCRIPTIONS = [
    'premium api access',
    'api access',
    'premium access',
    'access api',
    'api endpoint',
    'premium endpoint',
    'get data',
    'fetch data',
    'data api',
    'api service',
    'web service',
    'web api',
    'rest api',
    'http api',
];
export function isGenericDescription(desc) {
    if (!desc)
        return true;
    const trimmed = desc.trim();
    if (trimmed.length < 15)
        return true;
    const lower = trimmed.toLowerCase();
    return GENERIC_DESCRIPTIONS.some((g) => lower === g);
}
export function getDomainCounts(resources) {
    const counts = new Map();
    for (const r of resources) {
        try {
            const domain = new URL(r.resource).hostname;
            counts.set(domain, (counts.get(domain) ?? 0) + 1);
        }
        catch {
            // skip malformed URLs
        }
    }
    return counts;
}
export function scoreResourceQuality(resource, domainCounts) {
    let score = 0;
    // Output schema quality (weight: 0.30)
    const output = resource.metadata?.output;
    if (output && typeof output === 'object') {
        const outputObj = output;
        if (outputObj.output && typeof outputObj.output === 'object') {
            const inner = outputObj.output;
            if (inner.properties && typeof inner.properties === 'object') {
                score += 0.3;
            }
        }
    }
    // Description quality (weight: 0.30)
    const desc = resource.metadata?.description;
    if (desc && !isGenericDescription(desc) && desc.length > 50) {
        score += 0.3;
    }
    else if (desc && !isGenericDescription(desc)) {
        // Partial credit for non-generic but short descriptions
        score += 0.15;
    }
    // Domain concentration (weight: 0.20) — fewer resources = higher quality
    try {
        const domain = new URL(resource.resource).hostname;
        const count = domainCounts.get(domain) ?? 0;
        if (count < 100) {
            score += 0.2;
        }
        else if (count < 1000) {
            // Linear scale between 100 and 1000
            score += 0.2 * (1 - (count - 100) / 900);
        }
        // 1000+ gets 0
    }
    catch {
        // malformed URL gets 0
    }
    // HTTPS (weight: 0.10)
    if (resource.resource.startsWith('https://')) {
        score += 0.1;
    }
    // Timeout (weight: 0.10)
    const accepts = resource.accepts;
    if (accepts && accepts.length > 0) {
        // Check for maxTimeoutSeconds in accepts metadata
        const firstAccept = accepts[0];
        const timeout = firstAccept.maxTimeoutSeconds;
        if (timeout === undefined || timeout <= 30) {
            score += 0.1;
        }
        else if (timeout < 300) {
            // Linear scale between 30 and 300
            score += 0.1 * (1 - (timeout - 30) / 270);
        }
        // 300+ gets 0
    }
    else {
        // No timeout info — give benefit of the doubt
        score += 0.1;
    }
    return score;
}
export async function searchResources(query, resources, k = 5) {
    const candidateK = k * 5;
    const docs = resources.map((r) => new Document({
        pageContent: buildDocContent(r),
        metadata: { resource: r },
    }));
    const retriever = BM25Retriever.fromDocuments(docs, { k: candidateK });
    const results = await retriever.invoke(query);
    const candidates = results.map((doc) => doc.metadata.resource);
    const domainCounts = getDomainCounts(resources);
    const scored = candidates.map((resource, index) => {
        const qualityScore = scoreResourceQuality(resource, domainCounts);
        // BM25 position score: 1.0 for first result, decreasing linearly
        const bm25PositionScore = 1 - index / candidateK;
        const combinedScore = 0.4 * qualityScore + 0.6 * bm25PositionScore;
        return { resource, combinedScore };
    });
    scored.sort((a, b) => b.combinedScore - a.combinedScore);
    return scored.slice(0, k).map((s) => s.resource);
}
