// awal/dist/utils/errors.js
/**
 * Custom error classes for CLI operations
 */
export class WalletNotRunningError extends Error {
    constructor() {
        super('Wallet not running');
        this.name = 'WalletNotRunningError';
    }
    getDisplayMessage() {
        return `Wallet is not running.

The wallet server will start automatically when you run a command.
If it fails to start, you can run it manually with:
  npx awal show`;
    }
}
export class AuthRequiredError extends Error {
    constructor(message = 'Authentication required') {
        super(message);
        this.name = 'AuthRequiredError';
    }
    getDisplayMessage() {
        return `Authentication required.

Sign in using one of:
  1. Email OTP:
     npx awal auth login <your-email>
     npx awal auth verify <flow-id> <6-digit-code>

  2. Wallet UI:
     npx awal show`;
    }
}
export class IpcTimeoutError extends Error {
    constructor(channel, timeoutMs) {
        super(`IPC timeout on channel "${channel}" after ${timeoutMs}ms`);
        this.name = 'IpcTimeoutError';
    }
    getDisplayMessage() {
        return `Request timed out. The wallet may be unresponsive.

Try:
  1. Check if the wallet is still running: npx awal status
  2. Restart the wallet and try again`;
    }
}
export class IpcBridgeError extends Error {
    constructor(message) {
        super(message);
        this.name = 'IpcBridgeError';
    }
    getDisplayMessage() {
        return `Bridge communication error: ${this.message}

This may indicate a configuration issue. Try restarting the wallet.`;
    }
}
export function formatError(error) {
    if (error instanceof WalletNotRunningError) {
        return error.getDisplayMessage();
    }
    if (error instanceof AuthRequiredError) {
        return error.getDisplayMessage();
    }
    if (error instanceof IpcTimeoutError) {
        return error.getDisplayMessage();
    }
    if (error instanceof IpcBridgeError) {
        return error.getDisplayMessage();
    }
    if (error instanceof Error) {
        return `Error: ${error.message}`;
    }
    return `Unknown error: ${String(error)}`;
}
