// awal/dist/utils/serverManager.js
import envPaths from 'env-paths';
import { spawn, execSync } from 'child_process';
import { existsSync, mkdirSync, readFileSync, writeFileSync, cpSync } from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';
import { getPrimaryProcessPid } from './processCheck.js';
const paths = envPaths('awal');
// This will be updated by the build script
const BUNDLE_VERSION = '2.2.0';
function getServerPaths() {
    const dataDir = paths.data;
    const bundleDir = path.join(dataDir, 'server');
    // Electron binary location depends on platform
    const electronBinName = process.platform === 'win32' ? 'electron.cmd' : 'electron';
    return {
        dataDir,
        bundleDir,
        versionFile: path.join(bundleDir, '.version'),
        electronBin: path.join(bundleDir, 'node_modules', '.bin', electronBinName),
        bundleElectron: path.join(bundleDir, 'bundle-electron.js'),
    };
}
/**
 * Checks if the server bundle is installed and up to date
 */
async function isServerInstalled() {
    const { versionFile, electronBin, bundleElectron } = getServerPaths();
    // Check all required files exist
    if (!existsSync(versionFile) || !existsSync(electronBin) || !existsSync(bundleElectron)) {
        return false;
    }
    // Check version matches
    try {
        const installedVersion = readFileSync(versionFile, 'utf-8').trim();
        return installedVersion === BUNDLE_VERSION;
    }
    catch {
        return false;
    }
}
/**
 * Gets the path to the bundled server files within the npm package
 */
function getPackageBundleDir() {
    // This file is at dist/utils/serverManager.js
    // Server bundle is at server-bundle/
    const __dirname = path.dirname(fileURLToPath(import.meta.url));
    return path.join(__dirname, '..', '..', 'server-bundle');
}
/**
 * Installs the server bundle to the user's data directory
 */
async function installServer(onProgress) {
    const { bundleDir, versionFile, dataDir } = getServerPaths();
    const packageBundleDir = getPackageBundleDir();
    // Verify package bundle exists
    if (!existsSync(packageBundleDir)) {
        throw new Error(`Server bundle not found at ${packageBundleDir}. Package may be corrupted.`);
    }
    // Create data directory if needed
    if (!existsSync(dataDir)) {
        mkdirSync(dataDir, { recursive: true, mode: 0o700 });
    }
    // Copy bundled server files from package
    onProgress?.('Copying server bundle...');
    // Remove existing bundle if present
    if (existsSync(bundleDir)) {
        // Use rm -rf for cross-platform removal
        try {
            execSync(`rm -rf "${bundleDir}"`, { stdio: 'pipe' });
        }
        catch {
            // Fallback: try to overwrite
        }
    }
    cpSync(packageBundleDir, bundleDir, { recursive: true });
    // Install Electron (the heavy dependency)
    onProgress?.('Installing Electron runtime (one-time setup, may take a minute)...');
    await new Promise((resolve, reject) => {
        const npmCmd = process.platform === 'win32' ? 'npm.cmd' : 'npm';
        const child = spawn(npmCmd, ['install', '--omit=dev'], {
            cwd: bundleDir,
            stdio: 'inherit',
            shell: process.platform === 'win32',
        });
        child.on('error', (err) => reject(new Error(`npm install failed: ${err.message}`)));
        child.on('exit', (code) => {
            if (code === 0) {
                resolve();
            }
            else {
                reject(new Error(`npm install failed with exit code ${code}`));
            }
        });
    });
    // Write version marker
    writeFileSync(versionFile, BUNDLE_VERSION, { mode: 0o600 });
}
/**
 * Spawns the Electron server process in the background
 */
async function spawnServer() {
    const { electronBin, bundleElectron } = getServerPaths();
    // Verify electron exists
    if (!existsSync(electronBin)) {
        throw new Error(`Electron not found at ${electronBin}. Try reinstalling with: rm -rf ~/.local/share/awal && npx awal status`);
    }
    const child = spawn(electronBin, [bundleElectron], {
        detached: true,
        stdio: 'ignore',
        env: {
            ...process.env,
            STARTED_BY_CLI: 'true',
            WALLET_STANDALONE: 'true',
        },
    });
    // Don't wait for child, detach it
    child.unref();
    // Wait for server to initialize (poll for lock file)
    const maxWaitMs = 30000;
    const pollIntervalMs = 500;
    const startTime = Date.now();
    while (Date.now() - startTime < maxWaitMs) {
        await new Promise((r) => setTimeout(r, pollIntervalMs));
        if (getPrimaryProcessPid() !== null) {
            // Server is running, give it time to fully initialize
            // (Electron + React app + IPC handlers)
            await new Promise((r) => setTimeout(r, 8000));
            return;
        }
    }
    throw new Error('Server failed to start within timeout');
}
/**
 * Ensures the server is running, installing if needed
 * @returns true if server is running, false on failure
 */
export async function ensureServerRunning(onInstalling, onStarting) {
    // Check if already running
    if (getPrimaryProcessPid() !== null) {
        return true;
    }
    try {
        // Install if needed
        if (!(await isServerInstalled())) {
            onInstalling?.();
            await installServer((msg) => {
                // Progress messages are logged by the install process via stdio: 'inherit'
            });
        }
        // Start server
        onStarting?.();
        await spawnServer();
        return getPrimaryProcessPid() !== null;
    }
    catch (error) {
        console.error('Failed to start server:', error instanceof Error ? error.message : error);
        return false;
    }
}
/**
 * Checks if the wallet is running
 */
export function isWalletRunning() {
    return getPrimaryProcessPid() !== null;
}
