// awal/dist/commands/balance.js
import { Command } from 'commander';
import chalk from 'chalk';
import ora from 'ora';
import { sendIpcRequest } from '../ipcClient.js';
import { requireAuth } from '../utils/authCheck.js';
import { formatError } from '../utils/errors.js';
import { formatJson, formatAssetBalance } from '../formatters/output.js';
const EVM_DEFAULT_ASSETS = ['USDC', 'ETH', 'WETH'];
const SOLANA_DEFAULT_ASSETS = ['USDC', 'SOL'];
const SOLANA_CHAINS = ['solana', 'solana-devnet'];
function isSolanaChain(chain) {
    if (!chain)
        return false;
    return SOLANA_CHAINS.includes(chain) || chain.startsWith('solana:');
}
function getDefaultDecimals(asset) {
    const upper = asset.toUpperCase();
    if (upper === 'USDC')
        return 6;
    if (upper === 'SOL')
        return 9;
    return 18;
}
function printChainBalances(chainDisplay, assets, result) {
    console.log(chalk.bold(`\n${chainDisplay}`));
    console.log(chalk.dim('─'.repeat(24)));
    for (const asset of assets) {
        const assetData = result.balances?.[asset];
        const raw = assetData?.raw ?? '0';
        const decimals = assetData?.decimals ?? getDefaultDecimals(asset);
        const formatted = formatAssetBalance(asset, raw, decimals);
        console.log(`${chalk.cyan(asset.padEnd(8))}${formatted}`);
    }
}
export function createBalanceCommand() {
    return new Command('balance')
        .description('Get wallet balances across all chains (or a specific chain with --chain)')
        .option('--asset <asset>', 'Show specific asset only (usdc, eth, weth, sol)')
        .option('--chain <chain>', 'Show only one chain (base, base-sepolia, solana, solana-devnet)')
        .option('--json', 'Output as JSON')
        .action(async (options) => {
        const spinner = ora('Fetching balances...').start();
        try {
            await requireAuth();
            // Single-chain mode: explicit --chain flag
            if (options.chain) {
                const assets = options.asset
                    ? [options.asset.toUpperCase()]
                    : isSolanaChain(options.chain)
                        ? SOLANA_DEFAULT_ASSETS
                        : EVM_DEFAULT_ASSETS;
                const result = (await sendIpcRequest('get-wallet-balance', {
                    assets,
                    chain: options.chain,
                }));
                spinner.stop();
                if (options.json) {
                    console.log(formatJson(result));
                    return;
                }
                const chainDisplay = result.chain ?? options.chain;
                printChainBalances(chainDisplay, assets, result);
                return;
            }
            // All-chains mode: fetch EVM and Solana in parallel
            const evmAssets = options.asset ? [options.asset.toUpperCase()] : EVM_DEFAULT_ASSETS;
            const solanaAssets = options.asset ? [options.asset.toUpperCase()] : SOLANA_DEFAULT_ASSETS;
            const [evmResult, solanaResult] = (await Promise.all([
                sendIpcRequest('get-wallet-balance', { assets: evmAssets, chain: 'base' }),
                sendIpcRequest('get-wallet-balance', {
                    assets: solanaAssets,
                    chain: 'solana',
                }),
            ]));
            spinner.stop();
            if (options.json) {
                console.log(formatJson({ base: evmResult, solana: solanaResult }));
                return;
            }
            printChainBalances(evmResult.chain ?? 'Base', evmAssets, evmResult);
            printChainBalances(solanaResult.chain ?? 'Solana', solanaAssets, solanaResult);
        }
        catch (error) {
            spinner.fail('Failed to fetch balances');
            console.error(chalk.red(formatError(error)));
            process.exit(1);
        }
    });
}
