// awal/dist/commands/bazaar.js
import { Command } from 'commander';
import chalk from 'chalk';
import ora from 'ora';
import { sendIpcRequest } from '../ipcClient.js';
import { formatError } from '../utils/errors.js';
import { formatJson, formatTable, formatUsdcAmount } from '../formatters/output.js';
export function createBazaarCommand() {
    const bazaar = new Command('bazaar').description('Browse X402 payment-enabled services');
    bazaar
        .command('list')
        .description('List available bazaar resources')
        .option('--network <network>', 'Filter by network (base, base-sepolia)')
        .option('--full', 'Return complete details for all resources')
        .option('--json', 'Output as JSON')
        .action(async (options) => {
        const spinner = ora('Fetching bazaar resources...').start();
        try {
            const data = {};
            if (options.network) {
                data.network = options.network;
            }
            if (options.full) {
                data.full = true;
            }
            const result = (await sendIpcRequest('bazaar-list', data));
            spinner.stop();
            if (options.json) {
                console.log(formatJson(result));
                return;
            }
            // Table output
            if (!result.items || result.items.length === 0) {
                console.log('No resources found');
                return;
            }
            if (options.full) {
                // Full details - print each resource (handles nested accepts array)
                for (const item of result.items) {
                    const fullItem = item;
                    const accept = fullItem.accepts?.[0];
                    console.log(chalk.bold(`\n${item.resource}`));
                    if (accept?.description) {
                        console.log(chalk.dim(accept.description));
                    }
                    if (accept?.maxAmountRequired) {
                        console.log(`Price: ${formatUsdcAmount(parseInt(accept.maxAmountRequired, 10))}`);
                    }
                    if (accept?.network) {
                        console.log(`Network: ${accept.network}`);
                    }
                    if (accept?.scheme) {
                        console.log(`Scheme: ${accept.scheme}`);
                    }
                    if (fullItem.metadata?.confidence?.overallScore !== undefined) {
                        console.log(`Quality Score: ${fullItem.metadata.confidence.overallScore}`);
                    }
                    if (accept?.outputSchema) {
                        console.log(`Schema: ${JSON.stringify(accept.outputSchema, null, 2)}`);
                    }
                    console.log('---');
                }
            }
            else {
                // Summary table (uses flat structure from summary format)
                const tableData = result.items.map((r) => {
                    const summary = r;
                    return {
                        resource: summary.resource,
                        description: summary.description ?? '-',
                        price: summary.maxAmountRequired
                            ? formatUsdcAmount(parseInt(summary.maxAmountRequired, 10))
                            : '-',
                        network: summary.network ?? '-',
                    };
                });
                console.log(formatTable(tableData, [
                    { key: 'resource', header: 'Resource', width: 50 },
                    { key: 'description', header: 'Description', width: 40 },
                    { key: 'price', header: 'Price', width: 12 },
                    { key: 'network', header: 'Network', width: 12 },
                ]));
            }
        }
        catch (error) {
            spinner.fail('Failed to fetch bazaar resources');
            console.error(chalk.red(formatError(error)));
            process.exit(1);
        }
    });
    bazaar
        .command('search')
        .description('Search bazaar resources using BM25 relevance ranking')
        .argument('<query>', 'Search query')
        .option('--json', 'Output as JSON')
        .option('--network <network>', 'Filter by network')
        .option('-k, --top <n>', 'Number of results', '5')
        .action(async (query, options) => {
        const k = parseInt(options.top, 10);
        const spinner = ora('Searching...').start();
        try {
            const result = (await sendIpcRequest('bazaar-search', {
                query,
                limit: k,
                network: options.network,
            }));
            spinner.stop();
            const items = result.items || [];
            if (items.length === 0) {
                console.log('No matching resources found');
                return;
            }
            if (options.json) {
                console.log(formatJson(result));
                return;
            }
            console.log(chalk.dim(`Found ${items.length} result${items.length === 1 ? '' : 's'}\n`));
            for (const item of items) {
                const fullItem = item;
                console.log(chalk.bold(item.resource));
                const accept = fullItem.accepts?.[0];
                const description = fullItem.metadata?.description ?? accept?.description;
                if (description) {
                    console.log(chalk.dim(description));
                }
                if (accept?.maxAmountRequired) {
                    console.log(`Price: ${formatUsdcAmount(parseInt(accept.maxAmountRequired, 10))}`);
                }
                if (accept?.network) {
                    console.log(`Network: ${accept.network}`);
                }
                if (accept?.scheme) {
                    console.log(`Scheme: ${accept.scheme}`);
                }
                console.log('---');
            }
        }
        catch (error) {
            spinner.fail('Search failed');
            console.error(chalk.red(formatError(error)));
            process.exit(1);
        }
    });
    return bazaar;
}
