// awal/dist/commands/details.js
import { Command } from 'commander';
import chalk from 'chalk';
import ora from 'ora';
import { decodePaymentRequiredHeader } from '@x402/core/http';
import { formatError } from '../utils/errors.js';
import { formatJson } from '../formatters/output.js';
const HTTP_METHODS = ['GET', 'POST', 'PUT', 'DELETE', 'PATCH'];
function isEmpty(value) {
    if (value == null)
        return true;
    if (typeof value === 'string' && value === '')
        return true;
    if (Array.isArray(value) && value.length === 0)
        return true;
    if (typeof value === 'object' && Object.keys(value).length === 0)
        return true;
    return false;
}
function formatValue(value) {
    if (typeof value === 'object')
        return JSON.stringify(value, null, 2);
    return String(value);
}
function printFields(obj, indent) {
    for (const [key, value] of Object.entries(obj)) {
        if (isEmpty(value))
            continue;
        const formatted = formatValue(value);
        if (formatted.includes('\n')) {
            console.log(`${indent}${chalk.bold(key)}:`);
            for (const line of formatted.split('\n')) {
                console.log(`${indent}  ${line}`);
            }
        }
        else {
            console.log(`${indent}${chalk.bold(key)}: ${formatted}`);
        }
    }
}
export function createDetailsCommand() {
    return new Command('details')
        .description('Discover X402 payment requirements for an endpoint (auto-detects HTTP method)')
        .argument('<url>', 'Full URL of the X402-enabled endpoint')
        .option('--json', 'Output as JSON')
        .action(async (url, options) => {
        const spinner = ora('Discovering payment requirements...').start();
        try {
            let paymentRequired = null;
            let matchedMethod = null;
            for (const method of HTTP_METHODS) {
                try {
                    const response = await fetch(url, {
                        method,
                        headers: { Accept: 'application/json' },
                        redirect: 'follow',
                    });
                    if (response.status !== 402) {
                        continue;
                    }
                    // Try header-based decoding first (x402 v2)
                    const paymentRequiredHeader = response.headers.get('payment-required');
                    if (paymentRequiredHeader) {
                        try {
                            paymentRequired = decodePaymentRequiredHeader(paymentRequiredHeader);
                            matchedMethod = method;
                            break;
                        }
                        catch {
                            // Header decode failed, try body
                        }
                    }
                    // Fall back to body-based decoding (x402 v1 or plain JSON)
                    try {
                        const body = (await response.json());
                        if (body.accepts && Array.isArray(body.accepts)) {
                            paymentRequired = body;
                            matchedMethod = method;
                            break;
                        }
                    }
                    catch {
                        // Body parse failed, try next method
                    }
                }
                catch {
                    // Network error for this method, try next
                }
            }
            spinner.stop();
            if (!paymentRequired || !matchedMethod) {
                console.log(chalk.yellow('No X402 payment requirements found at that URL.'));
                process.exit(1);
            }
            if (options.json) {
                console.log(formatJson({ method: matchedMethod, url, ...paymentRequired }));
                return;
            }
            // Text output
            console.log(chalk.bold('Payment Requirements'));
            console.log('---');
            console.log(`URL: ${url}`);
            console.log(`Method: ${matchedMethod}`);
            // Print all non-empty top-level fields (except accepts, printed separately)
            const { accepts, resource, extensions, ...topLevel } = paymentRequired;
            printFields(topLevel, '');
            if (resource && typeof resource === 'object') {
                console.log(`\n${chalk.bold('Resource')}`);
                printFields(resource, '  ');
            }
            if (extensions && typeof extensions === 'object' && Object.keys(extensions).length > 0) {
                console.log(`\n${chalk.bold('Extensions')}`);
                printFields(extensions, '  ');
            }
            if (Array.isArray(accepts) && accepts.length > 0) {
                console.log(`\n${chalk.bold('Accepted Payment Options')}`);
                for (const req of accepts) {
                    console.log('');
                    printFields(req, '  ');
                }
            }
        }
        catch (error) {
            spinner.fail('Discovery failed');
            console.error(chalk.red(formatError(error)));
            process.exit(1);
        }
    });
}
