// awal/dist/commands/send.js
import { Command } from 'commander';
import chalk from 'chalk';
import ora from 'ora';
import { normalize } from 'viem/ens';
import { sendIpcRequest } from '../ipcClient.js';
import { requireAuth } from '../utils/authCheck.js';
import { formatError } from '../utils/errors.js';
import { formatJson, formatSuccess, formatUsdcAmount } from '../formatters/output.js';
import { isEthAddress, isEnsName, isSolanaAddress } from '../utils/addressValidation.js';
import { parseUsdcAmount } from '../utils/parseAmount.js';
import { getMainnetClient } from '../utils/viemClient.js';
function isSolanaChain(chain) {
    return chain === 'solana' || chain === 'solana-devnet';
}
export function createSendCommand() {
    return new Command('send')
        .description('Send USDC to an address')
        .argument('<amount>', 'Amount to send (e.g., "$0.01", "0.01", or atomic units)')
        .argument('<recipient>', 'Recipient address (0x... for EVM, Base58 for Solana) or ENS domain')
        .option('--chain <chain>', 'Blockchain network: base, base-sepolia, solana, solana-devnet (default: base)', 'base')
        .option('--json', 'Output as JSON')
        .action(async (amount, recipient, options) => {
        const spinner = ora('Preparing transaction...').start();
        try {
            // Check authentication first
            await requireAuth();
            // Parse amount to atomic units
            let atomicAmount;
            try {
                atomicAmount = parseUsdcAmount(amount);
            }
            catch (error) {
                spinner.fail('Invalid amount');
                console.error(chalk.red(formatError(error)));
                process.exit(1);
            }
            // Validate recipient address format
            const isSolana = isSolanaChain(options.chain);
            if (isSolana) {
                if (!isSolanaAddress(recipient)) {
                    spinner.fail('Invalid recipient');
                    console.error(chalk.red('Recipient must be a valid Solana address (Base58 encoded, 32-44 chars)'));
                    process.exit(1);
                }
            }
            else if (!isEthAddress(recipient) && !isEnsName(recipient)) {
                spinner.fail('Invalid recipient');
                console.error(chalk.red('Recipient must be a valid Ethereum address (0x...) or ENS domain'));
                process.exit(1);
            }
            let resolvedRecipient = recipient;
            let ensName;
            if (!isSolana && isEnsName(recipient)) {
                spinner.text = `Resolving ${recipient}...`;
                try {
                    const mainnetClient = getMainnetClient();
                    const address = await mainnetClient.getEnsAddress({
                        name: normalize(recipient),
                    });
                    if (!address) {
                        spinner.fail('ENS resolution failed');
                        console.error(chalk.red(`Could not resolve ENS name: ${recipient}. No address found.`));
                        process.exit(1);
                    }
                    resolvedRecipient = address;
                    ensName = recipient;
                }
                catch (error) {
                    spinner.fail('ENS resolution failed');
                    console.error(chalk.red(`Failed to resolve ENS name "${recipient}": ${formatError(error)}`));
                    process.exit(1);
                }
            }
            spinner.text = `Sending ${formatUsdcAmount(atomicAmount)} to ${ensName || resolvedRecipient}...`;
            const data = {
                recipient: resolvedRecipient,
                amountAtomic: atomicAmount,
                chain: options.chain,
            };
            const result = (await sendIpcRequest('send-transaction', data));
            spinner.stop();
            if (result.error) {
                console.error(chalk.red(`Transaction failed: ${result.error}`));
                process.exit(1);
            }
            if (options.json) {
                console.log(formatJson(result));
                return;
            }
            // Text output
            if (ensName) {
                console.log(formatSuccess(`Successfully sent ${formatUsdcAmount(atomicAmount)} to ${ensName}`));
                console.log(chalk.gray(`Resolved address: ${resolvedRecipient}`));
            }
            else {
                console.log(formatSuccess(`Successfully sent ${formatUsdcAmount(atomicAmount)} to ${recipient}`));
            }
            console.log(chalk.gray(`Chain: ${result.chain}`));
            if (result.transactionHash) {
                console.log(chalk.gray(`Transaction: ${result.transactionHash}`));
            }
            if (result.blockExplorerUrl) {
                const explorerLabel = isSolana ? 'View on Solscan' : 'View on Basescan';
                console.log(chalk.cyan(`\n${explorerLabel}: ${result.blockExplorerUrl}`));
            }
        }
        catch (error) {
            spinner.fail('Transaction failed');
            console.error(chalk.red(formatError(error)));
            process.exit(1);
        }
    });
}
