// awal/dist/commands/status.js
import { Command } from 'commander';
import chalk from 'chalk';
import ora from 'ora';
import { sendIpcRequest, setOnWalletStarting, setOnInstalling } from '../ipcClient.js';
import { getPrimaryProcessPid } from '../utils/processCheck.js';
import { formatError } from '../utils/errors.js';
import { formatJson, formatKeyValue, formatSuccess, formatWarning } from '../formatters/output.js';
export function createStatusCommand() {
    return new Command('status')
        .description('Check wallet server health and authentication status')
        .option('--json', 'Output as JSON')
        .action(async (options) => {
        const spinner = ora('Checking status...').start();
        // Check wallet server health first
        const walletPid = getPrimaryProcessPid();
        const serverRunning = walletPid !== null;
        if (!serverRunning) {
            spinner.text = 'Starting wallet...';
        }
        // Update spinner for installation and startup
        setOnInstalling(() => {
            spinner.text = 'Installing server bundle...';
        });
        setOnWalletStarting(() => {
            spinner.text = 'Starting wallet...';
        });
        try {
            const rawResult = await sendIpcRequest('check-session-status', {});
            spinner.stop();
            // Re-check after potential startup
            const currentPid = getPrimaryProcessPid();
            const serverHealthy = currentPid !== null;
            // The wallet returns a markdown string, parse it to determine auth status
            const resultString = typeof rawResult === 'string' ? rawResult : '';
            const isAuthenticated = resultString.includes('Authenticated') && !resultString.includes('Not Authenticated');
            // Extract email if present (format: "Logged in as email@example.com")
            const emailMatch = resultString.match(/Logged in as ([^\s]+)/);
            const email = emailMatch ? emailMatch[1] : undefined;
            if (options.json) {
                console.log(formatJson({
                    server: {
                        running: serverHealthy,
                        pid: currentPid,
                    },
                    auth: {
                        authenticated: isAuthenticated,
                        email,
                    },
                }));
                return;
            }
            // Text output - Server status
            console.log(chalk.bold('Wallet Server'));
            if (serverHealthy) {
                console.log(formatSuccess(`Running (PID: ${currentPid})`));
            }
            else {
                console.log(formatWarning('Not running'));
            }
            // Text output - Auth status
            console.log('');
            console.log(chalk.bold('Authentication'));
            if (isAuthenticated) {
                console.log(formatSuccess('Authenticated'));
                if (email) {
                    console.log(formatKeyValue({ 'Logged in as': email }));
                }
            }
            else {
                console.log(formatWarning('Not authenticated'));
                console.log(chalk.dim(`
Sign in using one of:
  1. Email OTP:
     npx awal auth login <your-email>
     npx awal auth verify <flow-id> <6-digit-code>

  2. Wallet UI:
     npx awal show`));
            }
        }
        catch (error) {
            spinner.fail('Failed to check status');
            console.error(chalk.red(formatError(error)));
            process.exit(1);
        }
    });
}
