// awal/dist/commands/trade.js
import { Command } from 'commander';
import chalk from 'chalk';
import ora from 'ora';
import { sendIpcRequest } from '../ipcClient.js';
import { requireAuth } from '../utils/authCheck.js';
import { formatError } from '../utils/errors.js';
import { formatJson, formatSuccess } from '../formatters/output.js';
import { isEthAddress } from '../utils/addressValidation.js';
// Known token aliases for display formatting
const TOKEN_ALIASES = {
    usdc: { decimals: 6, display: 'USDC' },
    eth: { decimals: 18, display: 'ETH' },
    weth: { decimals: 18, display: 'WETH' },
};
/**
 * Resolve a token for display purposes only.
 * The IPC handler will resolve addresses and decimals from chain.
 */
function resolveTokenDisplay(token) {
    const lower = token.toLowerCase();
    // Check if it's an alias
    if (TOKEN_ALIASES[lower]) {
        return {
            token: lower,
            decimals: TOKEN_ALIASES[lower].decimals,
            display: TOKEN_ALIASES[lower].display,
        };
    }
    // Check if it's a valid address
    if (isEthAddress(token)) {
        return {
            token: token,
            decimals: 18, // Display default, actual decimals fetched by IPC handler
            display: `${token.slice(0, 6)}...${token.slice(-4)}`,
        };
    }
    return null;
}
/**
 * Formats an amount for display (used for output messages)
 */
function formatAmountForDisplay(amount, display) {
    // Remove $ prefix for parsing
    const cleaned = amount.replace(/^\$/, '').trim();
    const num = parseFloat(cleaned);
    if (isNaN(num)) {
        return `${amount} ${display}`;
    }
    if (display === 'USDC') {
        return `$${num.toFixed(2)} USDC`;
    }
    return `${num} ${display}`;
}
/**
 * Formats atomic units as a human-readable amount for display
 */
function formatAtomicAmount(atomicUnits, decimals, display) {
    const num = Number(atomicUnits);
    const readable = num / 10 ** decimals;
    if (display === 'USDC') {
        return `$${readable.toFixed(2)} USDC`;
    }
    // For other tokens, show more precision
    return `${readable.toFixed(6)} ${display}`;
}
export function createTradeCommand() {
    return new Command('trade')
        .description('Swap tokens on Base network via CDP Swap API.\n\n' +
        'Examples:\n' +
        '  npx awal trade $1 usdc eth          # Swap $1 USDC for ETH\n' +
        '  npx awal trade 0.50 usdc eth        # Swap $0.50 USDC for ETH\n' +
        '  npx awal trade 0.01 eth usdc        # Swap 0.01 ETH for USDC\n' +
        '  npx awal trade 100 0x... 0x...      # Swap using contract addresses')
        .argument('<amount>', 'Amount to swap (e.g., "$1.00", "0.50", "500000")')
        .argument('<from>', 'Source token: alias (usdc, eth, weth) or contract address (0x...)')
        .argument('<to>', 'Destination token: alias (usdc, eth, weth) or contract address (0x...)')
        .option('-c, --chain <chain>', 'Blockchain network', 'base')
        .option('-s, --slippage <bps>', 'Slippage tolerance in basis points (100 = 1%)', '100')
        .option('--json', 'Output result as JSON')
        .action(async (amount, from, to, options) => {
        const spinner = ora('Preparing swap...').start();
        try {
            // Check authentication first
            await requireAuth();
            // Resolve tokens for display (IPC handler resolves addresses/decimals from chain)
            const fromResolved = resolveTokenDisplay(from);
            const toResolved = resolveTokenDisplay(to);
            if (!fromResolved) {
                spinner.fail('Invalid token');
                console.error(chalk.red(`Unknown token: ${from}. Use usdc, eth, weth, or a valid 0x address.`));
                process.exit(1);
            }
            if (!toResolved) {
                spinner.fail('Invalid token');
                console.error(chalk.red(`Unknown token: ${to}. Use usdc, eth, weth, or a valid 0x address.`));
                process.exit(1);
            }
            if (fromResolved.token.toLowerCase() === toResolved.token.toLowerCase()) {
                spinner.fail('Invalid swap');
                console.error(chalk.red('Cannot swap a token to itself.'));
                process.exit(1);
            }
            // Validate amount is a valid format
            const cleanedAmount = amount.replace(/^\$/, '').trim();
            if (!/^[\d.]+$/.test(cleanedAmount) || parseFloat(cleanedAmount) <= 0) {
                spinner.fail('Invalid amount');
                console.error(chalk.red(`Invalid amount: ${amount}. Must be a positive number.`));
                process.exit(1);
            }
            // Parse slippage
            const slippageBps = parseInt(options.slippage, 10);
            if (isNaN(slippageBps) || slippageBps < 0 || slippageBps > 5000) {
                spinner.fail('Invalid slippage');
                console.error(chalk.red('Slippage must be between 0 and 5000 basis points (0-50%)'));
                process.exit(1);
            }
            spinner.text = `Swapping ${formatAmountForDisplay(amount, fromResolved.display)} for ${toResolved.display}...`;
            // Pass token identifiers and human-readable amount to IPC handler
            // IPC handler will resolve addresses, fetch decimals, and convert to atomic
            const data = {
                fromToken: fromResolved.token,
                toToken: toResolved.token,
                fromAmount: amount, // Human-readable, IPC handler converts to atomic
                chain: options.chain,
                slippageBps,
            };
            const result = (await sendIpcRequest('execute-swap', data));
            spinner.stop();
            if (result.error || !result.success) {
                console.error(chalk.red(`Swap failed: ${result.error}`));
                process.exit(1);
            }
            if (options.json) {
                console.log(formatJson(result));
                return;
            }
            // Text output
            const fromFormatted = formatAmountForDisplay(amount, fromResolved.display);
            const toFormatted = result.toAmount
                ? formatAtomicAmount(result.toAmount, toResolved.decimals, toResolved.display)
                : toResolved.display;
            console.log(formatSuccess(`Successfully swapped ${fromFormatted} for ${toFormatted}`));
            console.log(chalk.gray(`Chain: ${result.chain}`));
            if (result.transactionHash) {
                console.log(chalk.gray(`Transaction: ${result.transactionHash}`));
            }
            if (result.blockExplorerUrl) {
                console.log(chalk.cyan(`\nView on Basescan: ${result.blockExplorerUrl}`));
            }
        }
        catch (error) {
            spinner.fail('Swap failed');
            console.error(chalk.red(formatError(error)));
            process.exit(1);
        }
    });
}
